sycl/doc/extensions/proposed/sycl_ext_oneapi_record_event.asciidoc

```diff
@@ -0,0 +1,219 @@
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension provides the ability to time the execution of commands in a
+queue without enabling profiling on the entire queue.
+This is more efficient on some platforms because only a subset of the events
+are required to contain timestamp information.
+It is also more convenient for use in libraries, where the library wants to
+get timing information for some commands, but the library does not control the
+construction of the queue (which is where the `enable_profiling` property is
+passed).
+
+This extension is structured as a free function, rather than a member function
+on `queue`, in order to be consistent with the API design in
+link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+sycl_ext_oneapi_enqueue_functions]
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RECORD_EVENT` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds the `ext_oneapi_queue_event_recording` enumerator to the
+`sycl::aspect` enumeration.
+
+```
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_queue_event_recording
+};
+
+} // namespace sycl
+```
+
+When a device has this aspect, the `record_event` function may be called for a
+queue on this device even if the queue is not constructed with the property
+`property::queue::enable_profiling`.
+
+=== New free function
+
+This extension adds the following free function.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+event record_event(const queue& q);
```

Comment:
If we do have this perhaps we could have the `queue` overload return `void` to match the other enqueue functions and use `submit_with_event` along with `barrier` to achieve these semantics.

```
event e = submit_with_event([&](handler &h) { barrier(h); });
```

This would be more verbose but I think it would be more consistent with the other enqueue functions.

Reply:
With the new name `submit_profiling_tag`, I think it is consistent to *not* have an overload that takes a `hander`.  The convention is that functions named "submit" only have a `queue` overload.

Reply:
So semantics-wise, I see the `submit`/`submit_with_events` functions as factories for connecting commands with dependencies, etc, rather than functions for creating/submitting commands themselves, so I think adding a new `submit*`  functions which create the command themselves, in this case an implicit `barrier` would still diverge from the current convention.

I think I would still prefer to use `submit_with_event` / `barrier` and a property for profiling (even though it would be a bit more verbose), as it would use the existing APIs and avoid introducing new ones, though I do prefer the new naming of `submit_profiling_tag` and I'm not strongly opposed to the approach of adding new `submit*` functions.

Reply:
> I think I would still prefer to use `submit_with_event` / `barrier` and a property for profiling (even though it would be a bit more verbose)

If you are not strongly opposed to the `submit_profiling_tag` name, I'd like to go with that.  We did consider exposing this functionality the way you propose, but concerns were raised about requiring people to call a function named "barrier" in order to get profiling information (even if the function actually performs a barrier).  Even if you don't agree with the naming argument, it does seem less verbose to introduce a new function here.